### PR TITLE
Remove need for inheritance in BindingDiagnosticBag

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/68041.  See description there for overall idea.

This PR removes the need for inheritance/overriding in the BindingDiagnosticBag hierarchy.  The only place where it is used to customize behavior between C# and VB now is supported by passing along a delegate with that specific functionality.  This will allow all existing behavior to now be provided with extensions, and eventually for the subclasses to go away.  

Full set of PRs are:

- [x] https://github.com/dotnet/roslyn/pull/68041
- [ ] https://github.com/dotnet/roslyn/pull/68042 (this PR)
- [ ] https://github.com/dotnet/roslyn/pull/68043
- [ ] https://github.com/dotnet/roslyn/pull/68044
- [ ] https://github.com/dotnet/roslyn/pull/68046

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@jjonescz @jaredpar this is ready for review.

File: src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs [hunk 1]
```diff
@@ -16,21 +16,23 @@ internal sealed class BindingDiagnosticBag : BindingDiagnosticBag<AssemblySymbol
     {
         public static readonly BindingDiagnosticBag Discarded = new BindingDiagnosticBag(null, null);
 
+        private static readonly Func<DiagnosticInfo, DiagnosticBag, Location, bool> s_reportUseSiteDiagnostic = Symbol.ReportUseSiteDiagnostic;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs:19]: Curious: do we need this pattern anymore? The compiler automatically caches method group conversions now.

**Review by jaredpar — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@jjonescz @jcouv ptal.

**Comment by CyrusNajmabadi** [comment]
@jjonescz ptal

**Review by jjonescz — APPROVED** [review]

**Pull request merged**